### PR TITLE
V6.2.0

### PR DESCRIPTION
**Pull request by KrisThielemans**

### CODE REVIEW

**Comment by KrisThielemans** [comment]
- Some Linux cuda jobs with python 3.12 [fail in initial Python import test](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=1006324&view=logs&j=3b077b59-19b3-5c3c-d9ea-4f2d359714ca&t=4427250c-9566-5227-2be0-fd328c273206&l=8463) with
  ```
  ImportError: libcudart.so.12: cannot open shared object file: No such file or directory
  ```
  Checking a few log files, it looks like these jobs installed `libparallelproj-cpu`, as opposed to `*cuda`. This leads to 2 questions:
  - Why do some cuda jobs ***not*** select the cuda version of `libparallelproj`. @gschramm any idea?
  - What is the STIR recipe missing to get `libcudart` if we wouldn't import `libparallelproj`? the `run` section does have
    ```
     - __cuda  # [cuda_compiler_version != "None"]
     ```
- Linux cuda jobs with python 3.11 or older (and hence install a `root_base` version that makes STIR switch to C++ 20) fail in CMake [with](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=1006324&view=logs&j=9a864fd9-6c8f-52ca-79ce-2aa6dca1a1de&t=10fc5aa2-324e-5982-4c88-6b31fcab16b3&l=1419)
  ```
  CMake Error in src/recon_buildblock/CMakeLists.txt:
  Target "recon_buildblock" requires the language dialect "CUDA20" .  But the
  current compiler "NVIDIA" does not support this, or CMake does not know the
  flags to enable it.
  ```
   which is the same error as in https://stackoverflow.com/a/75038504 where the answer says you need CMake 3.25.2 for C++ 20 support and CUDA, but the build is already using CMake 3.30.2.
- Linux cuda jobs that do not have the above problems [fail with](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=1006324&view=logs&j=bb1c2637-64c6-57bd-9ea6-93823b2df951&t=350df31b-3291-5209-0bb7-031395f0baa1&l=8687)
  ```
   Running OSMAPOSL CUDA/OSMAPOSL_test_sim_Parallelproj_CUDARDP.par 
   ./run_test_simulate_and_recon.sh: line 118:  4484 Segmentation fault 
  ```
  This is because there's no CUDA run-time on the azure infrastructure. We should therefore test for `nvidia-smi` here as well.
- osx jobs fail [with](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=1006324&view=logs&j=9c5ef928-2cd6-52e5-dbe6-9d173a7d951b&t=20c71c51-4b27-578b-485d-06ade2de1d00&l=2392)
  ```
  ld: library not found for -latomic
  ```
  these use clang. This will be due to the work-around I introduced in STIR for CLang OpenMP linking errors, see https://gitlab.kitware.com/cmake/cmake/-/issues/26037
- Windows CUDA jobs fail [with](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=1006324&view=logs&j=7df5a329-fcdb-5857-5720-a5dae94c0641&t=67fc0302-7ecf-51cb-2fd0-391a1258793e&l=9598)
  ```
  import stir
  File "D:\bld\stir_1724185476851\_test_env\lib\site-packages\stir.py", line 12, in <module>
    from _stir import *
  ImportError: DLL load failed while importing _stir: The specified module could not be found.
  ```
  which is likely the same problem as `libcudart.so not found`. The DLL checks above report
  ```
  WARNING (stir,Library/bin/extract_segments.exe): $RPATH/cudart64_12.dll not found in packages, sysroot(s) nor the missing_dso_whitelist.
  .. is this binary repackaging?
  ```
  However, I see this problem in at least one job that does install `libparallel-proj-cuda`, e.g. [here](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=1006324&view=logs&j=b36fd9fe-2c4c-5253-964c-f012f7496c0b&t=8537b377-2267-5ceb-9564-37d6e4aabbe8&l=9636)

**Comment by gschramm** [comment]
Hi Kris,

1. No idea why the builds with cuda comp v12 want to install the cpu version of libarallelproj :/
2. Regarding `libcudart`, the recipe for libparallelproj is below. There are extra `cuda-version` requirements for host and run, but I don't remember why we added those (with the help of the conda-forge people). Maybe those will solve the problems.
```
    requirements:
      build:
        - {{ stdlib("c") }}
        - {{ compiler('c') }}
        - {{ compiler('cxx') }}  # [cuda_compiler_version != "None"]
        - {{ compiler('cuda') }}  # [cuda_compiler_version != "None"]
        - cmake >=3.23
        - ninja
        - python
      host:
        - libgomp      # [linux]
        - llvm-openmp  # [osx]
        - cuda-version {{ cuda_compiler_version }}  # [(cuda_compiler_version or "None") != "None"]
      run:
        - __cuda  # [cuda_compiler_version != "None"]
        - cuda-version {{ cuda_compiler_version }}  # [(cuda_compiler_version or "None") != "None"]
```

**Comment by KrisThielemans** [comment]
> There are extra cuda-version requirements for host and run,

I did check that `cuda-version` is installed, so I think that won't help. but doing this check more carefully, I see in this job
- [`cuda_compiler_version:  '12.0'`](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=1006324&view=logs&j=3b077b59-19b3-5c3c-d9ea-4f2d359714ca&t=4427250c-9566-5227-2be0-fd328c273206&l=248)
- install of [12.6](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=1006324&view=logs&j=3b077b59-19b3-5c3c-d9ea-4f2d359714ca&t=4427250c-9566-5227-2be0-fd328c273206&l=8261)
- install of [12.0](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=1006324&view=logs&j=3b077b59-19b3-5c3c-d9ea-4f2d359714ca&t=4427250c-9566-5227-2be0-fd328c273206&l=756)
- [12.0 and `cuda-cudart-static` in the `build` section](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=1006324&view=logs&j=3b077b59-19b3-5c3c-d9ea-4f2d359714ca&t=4427250c-9566-5227-2be0-fd328c273206&l=837)
- [12.6 (and no cuda libraries) in the `host` section](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=1006324&view=logs&j=3b077b59-19b3-5c3c-d9ea-4f2d359714ca&t=4427250c-9566-5227-2be0-fd328c273206&l=869)
- [`cuda-version >=12.0,<13` and ` __cuda` in the `run` section](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=1006324&view=logs&j=3b077b59-19b3-5c3c-d9ea-4f2d359714ca&t=4427250c-9566-5227-2be0-fd328c273206&l=1030)
- Building success but the test run installs [`cuda-version` 12.6 but no cuda libraries](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=1006324&view=logs&j=3b077b59-19b3-5c3c-d9ea-4f2d359714ca&t=4427250c-9566-5227-2be0-fd328c273206&l=8261)

There's differences therefore between `host` and `build` in both version and installed libraries. As everything builds fine, I suspect that this is ok (and maybe we don't need `cuda` at all on the `host`). My feeling is that the problem is really just the last bullet. I suppose I could explicitly add `cuda-cudart` in the `run` section.... (I have no idea what `__cuda` adds).

**Comment by gschramm** [comment]
https://conda-forge.org/docs/maintainer/knowledge_base/#cuda-builds


> Conda exposes the maximum CUDA version supported by the installed Nvidia drivers through a virtual package named __cuda. By default, conda will install the highest version available for the packages involved. To override this behaviour, you can define a CONDA_OVERRIDE_CUDA environment variable. More details in the [Conda docs](https://docs.conda.io/projects/conda/en/stable/user-guide/tasks/manage-virtual.html#overriding-detected-packages).

**Comment by KrisThielemans** [comment]
@gchramm, why do you have
```
      ignore_run_exports:
        - cudatoolkit  # [(cuda_compiler_version or "None").startswith("11")]
```
No ignore for 12?

**Comment by KrisThielemans** [comment]
and https://github.com/conda-forge/cuda-feedstock/blob/main/recipe/doc/recipe_guide.md. Seems that the `cuda-version` might be a good idea after all. There's far too much there for me to spend time on though. Sigh.

**Comment by gschramm** [comment]
Indeed challenging to keep up to date. That's why I usually ask for help on the conda-forge gitter (which usually works very well)

**Comment by KrisThielemans** [comment]
@conda-forge/core I hope you can help. The main change here is that STIR now has CUDA code itself, so we need to take that into account. Summarising above problems related to CUDA:
- For CUDA12 jobs,  `libcudart.so.12` (or DLL) is missing for the test, so the "import stir` test fails. Do we need an explicit dependency?
- The hope was that if CUDA is detected, `libparallelproj` would choose its CUDA version as well, but that doesn't seem to be guaranteed.

Obviously, your help with the CMake issue would also be appreciated, but I'm not expecting any.

**Comment by hmaarrfk** [comment]
See https://github.com/conda-forge/pytorch-cpu-feedstock/blob/main/recipe/meta.yaml for an example of how to build cuda recipes.

**Comment by KrisThielemans** [comment]
- Linux CUDA 12 jobs now work (older CUDA versoins fail due to unrelated CMake C++-20 problem).
- Windows CUDA 12 fail with `cuda-driver-dev does not exist`.  Older CUDA version still fail with DLL not found. That's not a surprise as the previous fix was only for CUDA 12, but I have no idea what to do for CUDA 11. Anyone?
- OSX jobs still fail due to STIR CMake clang problem mentioned above

**Comment by hmaarrfk** [comment]
> but I have no idea what to do for CUDA 11. Anyone?

its probably fine to drop in 2024 at the feedstock level. don't let it hold you back. 

If somebody wants it back they can make a PR and bring back support.

**Comment by hmaarrfk** [comment]
I think this can be used as an example for windows maybe https://github.com/conda-forge/pytorch-cpu-feedstock/pull/231

**Comment by KrisThielemans** [comment]
@conda-forge-admin please rerender

**Comment by KrisThielemans** [comment]
@conda-forge-admin please rerender

**Comment by KrisThielemans** [comment]
@conda-forge-admin please rerender

**Comment by KrisThielemans** [comment]
Seems that I know skip everything and can't get rid of it. @casperdcl any ideas?

**Comment by KrisThielemans** [comment]
@conda-forge-admin, please lint

**Comment by gschramm** [comment]
@KrisThielemans : When I look at the "Files Changed" tab of this PR, it seems that indeed all the files .ci_support files have been deleted (not sure if that was done by any of the bots)

**Comment by KrisThielemans** [comment]
yup, the rerender. I'm a bit confused by the sequence here. I had some white-space issues in the recipe etc, but I suppose I just have the wrong `skip` clause (trying to do only cuda 12 on Windows, and skipping osx for now)

File: recipe/meta.yaml [hunk 1]
```diff
@@ -1,22 +1,32 @@
...
 
 source:
   url: https://github.com/UCL/STIR/archive/refs/tags/rel_{{ version }}.tar.gz
-  sha256: 638121f9b3c21e88cc1b6e6f66d8a89ed2902f9a84a546dfe0b373ea9aa5ce11
+  sha256: c2c072b50545a16484951a6412ab5bbb2944320f5c7de12ead7329725d3d7044
 
 build:
   number: {{ build_number }}
-  skip: true  # [linux and cuda_compiler_version == "10.2"]
+  skip: true
```
Comment (h-vetinari) [inline recipe/meta.yaml:25]:
```suggestion
```

**Comment by h-vetinari** [comment]
@conda-forge-admin, please rerender

File: recipe/meta.yaml [hunk 2]
```diff
@@ -1,22 +1,31 @@
...
 source:
   url: https://github.com/UCL/STIR/archive/refs/tags/rel_{{ version }}.tar.gz
-  sha256: 638121f9b3c21e88cc1b6e6f66d8a89ed2902f9a84a546dfe0b373ea9aa5ce11
+  sha256: c2c072b50545a16484951a6412ab5bbb2944320f5c7de12ead7329725d3d7044
 
 build:
   number: {{ build_number }}
-  skip: true  # [linux and cuda_compiler_version == "10.2"]
+  # skipping CUDA 11 builds on Windows due to missing cudart.dll problems, see https://github.com/conda-forge/stir-feedstock/pull/121
+  skip: true  # [win and (cuda_major<12)]
```
Comment (h-vetinari) [inline recipe/meta.yaml:26]: This probably doesn't work out with the different parsing phases (jinja vs. selectors etc.).
```suggestion
  {% if cuda_major < 12 %}
  skip: true  # [win]
  {% endif %}
```

File: recipe/meta.yaml [hunk 3]
```diff
@@ -44,6 +57,11 @@ requirements:
...
     - libitk
     - nlohmann_json
+    - cuda-version {{ cuda_compiler_version }}  # [(cuda_compiler_version or "None") != "None"]
+    {% if cuda_major >= 12 %}
+    - cuda-driver-dev                        # [linux]
+    - cuda-cudart-dev
+    {% endif %}
     - libparallelproj
     # osx-arm skipped due to https://github.com/conda-forge/root-feedstock/issues/260
     - root_base  # [not (py==312 or win or (osx and arm64))]
```
Comment (KrisThielemans) [inline recipe/meta.yaml:67]:
```suggestion
    # Recent ROOT needs C++ 20, but CUDA 11 doesn't support it
    # osx-arm skipped due to https://github.com/conda-forge/root-feedstock/issues/260
    # Python 3.12 not yet supported in ROOT
    {% if cuda_major < 12 %}
    - root_base  # [not (py==312 or win or (osx and arm64))]
    {% endif %}
```

**Comment by KrisThielemans** [comment]
@conda-forge-admin, please rerender

**Comment by KrisThielemans** [comment]
Thanks @h-vetinari for fixing my (stupid) mistake with the "bare" `skip` and the suggestion. As always, great support from the core!

Hopefully we're there now.

File: recipe/meta.yaml [hunk 4]
```diff
@@ -44,14 +59,24 @@ requirements:
...
+    - cuda-version {{ cuda_compiler_version }}  # [(cuda_compiler_version or "None") != "None"]
+    {% if cuda_major >= 12 %}
+    - cuda-driver-dev                        # [linux]
+    - cuda-cudart-dev
+    {% endif %}
     - libparallelproj
+    # Recent ROOT needs C++ 20, but CUDA 11 doesn't support it
     # osx-arm skipped due to https://github.com/conda-forge/root-feedstock/issues/260
+    # Python 3.12 not yet supported in ROOT
+    {% if cuda_major < 12 %}
```
Comment (KrisThielemans) [inline recipe/meta.yaml:71]:
```suggestion
    {% if cuda_major != 11 %}
```

File: recipe/meta.yaml [hunk 5]
```diff
@@ -1,22 +1,33 @@
...
 source:
   url: https://github.com/UCL/STIR/archive/refs/tags/rel_{{ version }}.tar.gz
-  sha256: 638121f9b3c21e88cc1b6e6f66d8a89ed2902f9a84a546dfe0b373ea9aa5ce11
+  sha256: c2c072b50545a16484951a6412ab5bbb2944320f5c7de12ead7329725d3d7044
 
 build:
   number: {{ build_number }}
-  skip: true  # [linux and cuda_compiler_version == "10.2"]
+  # skipping CUDA 11 builds on Windows due to missing cudart.dll problems, see https://github.com/conda-forge/stir-feedstock/pull/121
+  {% if cuda_major < 12 %}
```
Comment (KrisThielemans) [inline recipe/meta.yaml:26]:
```suggestion
  {% if cuda_major == 11 %}
```

**Comment by KrisThielemans** [comment]
@conda-forge-admin, please rerender

**Pull request merged**